commands/general.go

```diff
@@ -0,0 +1,54 @@
+package commands
+
+import (
+	"net/http"
+
+	"github.com/openfaas/faas-cli/config"
+	"github.com/openfaas/faas-cli/proxy"
+)
+
+//CLIAuth auth struct for the CLI
+type CLIAuth struct {
```

Comment:
I would create two types,

```go
type BasicAuth struct {
    Username string
    password string
}

func (c *BasicAuth) Set(req *http.Request) error {
    req.SetBasicAuth(c.Username, c.Password)
    return nil
}

type BearerToken struct {
     token string
}

func (c *BearerToken) Set(req *http.Request) error {
    req.Header.Set("Authorization", "Bearer "+c.token)
    return nil
}
```

This means each type only implements it's own logic and both satisfy the interface.   `NewCLIAuth` can just return pointers `&BearerToken{token: token}` or `&BasicAuth{Username: username, password: password` based on the config.

Reply:
Is this required for the PR to be merged? The code seems relatively straightforward as is

Reply:
I have read that article and agree with the point and sentiment, but what is the point of an Interface when the only implementation is a super object that implements both versions of authentication. I am not suggesting that it be rewitten to somehow make it more flexible, generic, or indirect.  Instead, to create two smaller focused implementations. Right now we have one auth object that implements two different flows.

If we decide to move this code into SDK, then would say it would be important to split it and make the implementations clearly named and focused. But as it stands, being part of the CLI and not being part of the SDK ...  it is fine to leave it

Reply:
I have done this change.